humility-core/src/arch_rv.rs

```diff
@@ -0,0 +1,322 @@
+                RegisterField::bit(3, "nmip"),
+                RegisterField::bit(2, "step"),
+                RegisterField::field(1, 0, "priv"),
+            ]),
+            RVRegister::MTVEC => Some(vec![RegisterField::field(1, 0, "mode")]),
+            _ => None,
+        }
+    }
+}
+
+use capstone::prelude::*;
+
+impl From<RegId> for RVRegister {
+    fn from(reg: RegId) -> Self {
+        use arch::riscv::RiscVReg::*;
+
+        match reg.0 as u32 {
+            RISCV_REG_ZERO => RVRegister::ZERO,
+            RISCV_REG_RA => RVRegister::RA,
+            RISCV_REG_SP => RVRegister::SP,
+            RISCV_REG_GP => RVRegister::TP,
+            RISCV_REG_T0 => RVRegister::T0,
+            RISCV_REG_T1 => RVRegister::T1,
+            RISCV_REG_T2 => RVRegister::T2,
+            RISCV_REG_T3 => RVRegister::T3,
+            RISCV_REG_T4 => RVRegister::T4,
+            RISCV_REG_T5 => RVRegister::T5,
+            RISCV_REG_T6 => RVRegister::T6,
+            RISCV_REG_TP => RVRegister::TP,
+            RISCV_REG_S0 => RVRegister::S0,
+            RISCV_REG_S1 => RVRegister::S1,
+            RISCV_REG_S2 => RVRegister::S2,
+            RISCV_REG_S3 => RVRegister::S3,
+            RISCV_REG_S4 => RVRegister::S4,
+            RISCV_REG_S5 => RVRegister::S5,
+            RISCV_REG_S6 => RVRegister::S6,
+            RISCV_REG_S7 => RVRegister::S7,
+            RISCV_REG_S8 => RVRegister::S8,
+            RISCV_REG_S9 => RVRegister::S9,
+            RISCV_REG_S10 => RVRegister::S10,
+            RISCV_REG_S11 => RVRegister::S11,
+            _ => {
+                panic!("unrecognized register {:x}", reg.0);
+            }
+        }
+    }
+}
+
+impl std::fmt::Display for RVRegister {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        formatter.pad(&format!("RV: {:?}", self))
+    }
+}
+
+//
+// our stub frames (that is, those frames that contain system call
+// instructions) have no DWARF information that describes how to unwind
+// through them; for these frames we do some (very crude) analysis of the
+// program text to determine what registers are pushed and how they are
+// manipulated so we can properly determine register state before the system
+// call. This is currently incomplete as it assums the registers are stored in order
+// TODO the conditions to push probably need more rigor
+pub fn riscv_presyscall_pushes(
+    cs: &Capstone,
+    instrs: &[capstone::Insn],
+) -> Result<Vec<Register>> {
+    const RV_INSN_SW: u32 = arch::riscv::RiscVInsn::RISCV_INS_SW as u32;
+    const RV_INSN_C_SW: u32 = arch::riscv::RiscVInsn::RISCV_INS_C_SW as u32;
+
+    let mut rval = vec![];
+    for instr in instrs {
+        match instr.id() {
+            InsnId(RV_INSN_C_SW) | InsnId(RV_INSN_SW) => {
+                for op in instr_operands(cs, instr).iter().rev() {
+                    rval.push(*op);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    Ok(rval)
```

Comment:
Don't we need a `rval.reverse();` here like arm does?

Reply:
probably... This does not actually get called yet as stack unwinding is broken.  I added `.reverse()` and updated the comment to reflect this.